src/core/gpio_monitor.c

```diff
@@ -2,209 +2,312 @@
         (void)PORTB;
         CNENBSET = _CNENB_CNIEB8_MASK;
-        callback_id = PWM_CALLBACK;
         break;
-        case GPIO_INT:
+    case GPIO_INT:
         CNCONDSET = _CNCOND_ON_MASK;
         IEC1SET = _IEC1_CNDIE_MASK;
         IFS1CLR = _IFS1_CNDIF_MASK;
         (void)PORTD;
         CNENDSET = _CNEND_CNIED0_MASK;
-        callback_id = INT_CALLBACK;
         break;
-        case GPIO_RST:
+    case GPIO_RST:
         CNCONDSET = _CNCOND_ON_MASK;
         IEC1SET = _IEC1_CNDIE_MASK;
         IFS1CLR = _IFS1_CNDIF_MASK;
         (void)PORTD;
         CNENDSET = _CNEND_CNIED6_MASK;
-        callback_id = RST_CALLBACK;
         break;
-        case GPIO_CS:
+    case GPIO_CS:
         CNCONESET = _CNCONE_ON_MASK;
         IEC1SET = _IEC1_CNEIE_MASK;
         IFS1CLR = _IFS1_CNEIF_MASK;
         (void)PORTE;
         CNENESET = _CNENE_CNIEE5_MASK;
-        callback_id = CS_CALLBACK;
         break;
-        default:
+    default:
         fprintf(stderr, "gpio_monitor: Invalid pin");
         return -1;
     }
 
-    callbacks[callback_id].event_mask = event_mask;
-    callbacks[callback_id].callback = callback;
-    return callback_id;
+    return 0;
 }
 
-int gpio_monitor_remove_callback(int callback_ID)
+static int disable_interrupt(uint8_t gpio_pin)
 {
-    if(callback_ID < 0
-        || callback_ID >= CALLBACK_COUNT
-        || callbacks[callback_ID].callback == NULL)
-    {
-        fprintf(stderr, "gpio_monitor: Invalid callback id");
-        return -1;
-    }
-
-    switch(callback_ID) {
-        case AN_CALLBACK:
+    switch(gpio_pin) {
+    case GPIO_AN:
         CNCONGCLR = _CNCONG_ON_MASK;
         IEC1CLR = _IEC1_CNGIE_MASK;
         IFS1CLR = _IFS1_CNGIF_MASK;
         (void)PORTG;
         CNENGCLR = _CNENG_CNIEG9_MASK;
         break;
-        case PWM_CALLBACK:
+    case GPIO_PWM:
         CNCONBCLR = _CNCONB_ON_MASK;
         IEC1CLR = _IEC1_CNBIE_MASK;
         IFS1CLR = _IFS1_CNBIF_MASK;
         (void)PORTB;
         CNENBCLR = _CNENB_CNIEB8_MASK;
         break;
-        case INT_CALLBACK:
+    case GPIO_INT:
         CNCONDCLR = _CNCOND_ON_MASK;
         IEC1CLR = _IEC1_CNDIE_MASK;
         IFS1CLR = _IFS1_CNDIF_MASK;
         (void)PORTD;
         CNENDCLR = _CNEND_CNIED0_MASK;
         break;
-        case RST_CALLBACK:
+   case GPIO_RST:
         CNCONDCLR = _CNCOND_ON_MASK;
         IEC1CLR = _IEC1_CNDIE_MASK;
         IFS1CLR = _IFS1_CNDIF_MASK;
         (void)PORTD;
         CNENDCLR = _CNEND_CNIED6_MASK;
         break;
-        case CS_CALLBACK:
+    case GPIO_CS:
         CNCONECLR = _CNCONE_ON_MASK;
         IEC1CLR = _IEC1_CNEIE_MASK;
         IFS1CLR = _IFS1_CNEIF_MASK;
         (void)PORTE;
         CNENECLR = _CNENE_CNIEE5_MASK;
         break;
-        default:
-        fprintf(stderr, "gpio_monitor: Invalid callback id\n");
+    default:
+        fprintf(stderr, "gpio_monitor: Invalid pin\n");
+        return -1;
+    }
+
+    return 0;
+}
+
+static int convert_gpio_pin(uint8_t gpio_pin)
```

Comment:
Wouldn't it make more sense to make this a function that's part of gpio.h that lets you to convert gpio_pins back to their types? I.e. get_gpio_type(uint8_t gpio_pin)?